paltas/Analysis/conv_models.py

```diff
@@ -182,17 +177,111 @@ def build_xresnet34(img_size,num_outputs,custom_head=False,
-		x = layers.Dense(num_outputs,use_bias=False,name='head_dense2')(x)
+			name='head_bn2',trainable=output_trainable)(x)
+		x = layers.Dense(num_outputs,use_bias=False,name='head_dense2',
+			trainable=output_trainable)(x)
 		outputs = layers.BatchNormalization(axis=bn_axis,epsilon=1e-5,
-			momentum=0.1,name='head_bn3')(x)
+			momentum=0.1,name='head_bn3',trainable=output_trainable)(x)
 	else:
-		outputs = layers.Dense(num_outputs,name='output_dense')(x)
+		outputs = layers.Dense(num_outputs,name='output_dense',
+			trainable=output_trainable)(x)
+
+	return outputs
+
+
+def build_xresnet34(img_size,num_outputs,custom_head=False,
+	train_only_head=False):
+	""" Build the xresnet34 model described in
+	https://arxiv.org/pdf/1812.01187.pdf
+
+	Args:
+		img_size ((int,int,int)): A tuple with shape (pix,pix,freq) that
+			describes the size of the input images.
+		num_outputs (int): The number of outputs to predict
+		custom_head (bool): If true, then add a custom head at the end of
+			xresnet34 in line with what' used in the fastai code.
+		train_only_head (bool): If true, only train the head of the network.
+
+	Returns:
+		(keras.Model): An instance of the xresnet34 model implemented in
+		Keras.
+	"""
+
+	# If we train only the head, then none of the previous weights should be
+	# trainable
+	trainable = not train_only_head
+
+	# Initialize the inputs
+	inputs = layers.Input(shape=img_size)
+
+	# Pass the inputs through out xresnet 34 model.
+	outputs = _xresnet34(inputs,num_outputs,custom_head=custom_head,
+		trainable=trainable,output_trainable=True)
 
 	model = Model(inputs=inputs,outputs=outputs)
 
 	return model
+
+
+def build_xresnet34_fc_inputs(img_size,num_outputs,num_fc_inputs,
+	train_only_head=False):
+	""" Build the xresnet34 model described in
+	https://arxiv.org/pdf/1812.01187.pdf but include a few more
+	fully connected layers and introduce additional inputs (i.e.
+	float values associated with the images) into the fully
+	connected layers.
+
+	Args:
+		img_size ((int,int,int)): A tuple with shape (pix,pix,freq) that
+			describes the size of the input images.
+		num_outputs (int): The number of outputs to predict
+		num_fc_inputs (int): The number of input floats associated to each
+			image.
+		train_only_head (bool): If true, only train the head of the network.
+
+	Returns:
+		(keras.Model): An instance of the xresnet34 model implemented in
+		Keras.
+	"""
+
+	# If we train only the head, then none of the previous weights should be
+	# trainable
+	trainable = not train_only_head
+
+	# Initialize the image inputs
+	inputs_image = layers.Input(shape=img_size)
+
+	# Initialize the inputs to the fully connected layer.
+	inputs_fc = layers.Input(shape=(num_fc_inputs,))
+
+	# The output of our xresnet is not the input to our fc stack
+	xresnet_output = 512
+
+	# Pass the inputs through out xresnet 34 model.
+	xr_outputs = _xresnet34(inputs_image,xresnet_output,custom_head=False,
+		trainable=trainable,output_trainable=trainable)
```

Comment:
Here I mean the fully connected layers after the redshifts are passed in to be the "head," so the "output" of _xresnet34 should only be trained if we want to train the body. However, I have changed the layers that get added after the redshift is concatenated so that they are always trainable.